spec/ics-25-handler-interface/README.md

```diff
@@ -0,0 +1,366 @@
+---
+
+## Synopsis
+
+This document describes the interface exposed by the standard IBC implementation (referred to as the IBC handler) to modules within the same state machine, and the implementation of that interface by the IBC handler.
+
+### Motivation
+
+IBC is an inter-module communication protocol, designed to faciliate reliable, authentication message passing between modules on separate blockchains. Modules should be able to reason about the interface they interact with and the requirements they must adhere to in order to utilize it safely.
+
+### Definitions
+
+`Client` and `ConsensusState` are as defined in ICS 2.
+
+`Connection` and `ConnectionState` are as defined in ICS 3.
+
+`Channel` and `Packet` are as defined in ICS 4.
+
+`CommitmentProof` is as defined in ICS 23.
+
+`Identifier`s must conform to the schema defined in ICS 24.
+
+### Desired Properties
+
+- Creation of clients, connections, and channels should be as permissionless as possible.
+- The module set should be dynamic: chains should be able to add and destroy modules at will with a persistent IBC handler.
+- Modules should be able to write their own more complex abstractions on top of IBC to provide additional semantics or guarantees.
+
+## Specification
+
+### Clients
+
+By default, clients are unowned: any module can create a new client, query any existing client, update any existing client, and delete any existing client not in use.
+
+`ClientKind` enumerates the list of light client algorithms supported by the handler implementation.
+
+```golang
+type ClientKind enum {
+  Tendermint
+}
+```
+
+`ClientOptions` contains all the parameter choices required to create a client.
+
+```golang
+type ClientOptions struct {
+  ClientKind      kind
+  ConsensusState  consensusState
+}
+```
+
+`ClientInfo` contains information about an existing client.
+
+```golang
+type ClientInfo struct {
+  ClientKind      kind
+  ConsensusState  consensusState
+}
+```
+
+`createClient` creates a new client and returns an automatically allocated identifier.
+
+```coffeescript
+function createClient(ClientOptions options) -> string
+```
+
+`queryClient` queries a client by a known identifier, returning the associated metadata and consensus state if found.
+
+```coffeescript
+function queryClient(string identifier) -> Maybe<ClientInfo>
+```
+
+`updateClient` updates an existing client with a new header, returning an error if the client was not found or the header was not a valid update.
+
+The default IBC relayer module will allow external calls to `updateClient`.
+
+```coffeescript
+function updateClient(string identifier, Header header) -> Maybe<Err>
+```
+
+`freezeClient` freezes an existing client by providing proof-of-equivocation, automatically freezing any associated connections & channels.
+
+The default IBC relayer module will allow external calls to `freezeClient`.
+
+```coffeescript
+function freezeClient(string identifier, Header headerOne, Header headerTwo) -> Maybe<Err>
```

Comment:
Is the equivocation proof verification should be done inside the `freezeClient`? As far as I understand the handler interface is exposed to the internal logic only, not the external accounts, so the application logic can decide when to call this function. By separating proving and freezing we can handle other cases to freeze the client in the future easily. 

In the implementation side, it will be a bit tricky to verify the equivocation inside this function, as it requires to read past validator set data, which I think better to be separated from the client logic.